NYPLAudiobookToolkit/DRM/FeedbookDRMProcessor.swift

```diff
@@ -0,0 +1,139 @@
+import CommonCrypto
+
+class FeedbookDRMProcessor {
+    class func processManifest(_ manifest: [String: Any], drmData: inout [String: Any]) -> Bool {
+        guard let metadata = manifest["metadata"] as? [String: Any] else {
+            ATLog(.info, "[FeedbookDRMProcessor] no metadata in manifest")
+            return true
+        }
+        
+        // Perform Feedbooks DRM rights check
+        if let feedbooksRights = metadata["http://www.feedbooks.com/audiobooks/rights"] as? [String: Any] {
+            if let startDate = DateUtils.parseDate((feedbooksRights["start"] as? String) ?? "") {
+                if Date() < startDate {
+                    ATLog(.error, "Feedbook DRM rights start date is in the future!")
+                    return false
+                }
+            }
+            if let endDate = DateUtils.parseDate((feedbooksRights["end"] as? String) ?? "") {
+                if Date() > endDate {
+                    ATLog(.error, "Feedbook DRM rights end date is expired!")
+                    return false
+                }
+            }
+        }
+        
+        // Perform Feedbooks DRM license status check
+        let licenseCheckUrl: URL?
+        if let links = manifest["links"] as? [[String: Any]] {
+            var href = ""
+            var found = false
+            for link in links {
+                if (link["rel"] as? String) == "license" {
+                    if found {
+                        ATLog(.warn, "[Feedbook License Status Check] More than one license status link found?! href:\(link["href"] ?? "") type:\(link["type"] ?? "")")
+                        continue
+                    }
+                    found = true
+                    href = (link["href"] as? String) ?? ""
+                }
+            }
+            licenseCheckUrl = URL(string: href)
+            drmData["status"] = DrmStatus.processing
+        } else {
+            licenseCheckUrl = nil
+        }
+        if licenseCheckUrl != nil {
+            drmData["licenseCheckUrl"] = licenseCheckUrl!
+        }
+        
+        // Perform Feedbooks manifest validation
+        // TODO:
+        
+        return true
+    }
+    
+    class func performAsyncDrm(book: OpenAccessAudiobook, drmData: [String: Any]) {
+        if let licenseCheckUrl = drmData["licenseCheckUrl"] as? URL {
+            weak var weakBook = book
+            URLSession.shared.dataTask(with: licenseCheckUrl) { (data, response, error) in
+                // Errors automatically mean success
+                if error != nil {
+                    weakBook?.drmStatus = .succeeded
```

Comment:
I hope this clarifies it?

Reply:
thanks, it does but... why are we doing the license check anyway then? all the user has to do is to turn off internet to skip it.

Reply:
That's a fair point. Perhaps @leonardr can offer an opinion here.

Reply:
This is a general feature of DRM rules: if a network check fails then it's okay to assume the check succeeded rather than it failed. It's spelled out explicitly in [the LCP spec](https://readium.org/lcp-specs/releases/lcp/latest#74-signature-processing), which Feedbooks audio DRM is based on: "Reading Systems must not... Block a User from opening a Publication if a network connection is unavailable"

All DRM decisions are based on a trade-off between strictness and usability, and this is the main one where the decision comes down on the side of usability. Mainly because the worst that will happen is someone keeps a book they are authorized to read for longer than the loan period.

Reply:
Sounds good. For DRM systems I previously worked on that was a not allowed, so I was confused. All good here then.